File: rs/src/katna.rs
//! Functions for decomposing a lujvo into **rafsi**.

use itertools::Itertools as _;

use crate::{
    data::{HYPHENS, INITIAL, MZ_VALID, VALID},
    exceptions::Jvonunfli::{
        self, DecompositionError, FakeTypeError, InvalidClusterError, NoLujvoFoundError,
        NotBrivlaError,
    },
    extract,
    jvozba::{get_lujvo_from_list, score, tiebreak},
    rafsi::RAFSI,
    strin, strsl,
    tarmi::{
        BrivlaType::{Cmevla, ExtendedLujvo, Lujvo},
        Settings,
        Tarmi::{Ccv, Ccvc, Ccvcv, Cvc, Cvcc, Cvccv, Cvhv, Cvv},
        YHyphenSetting::{ForceY, Standard},
        is_consonant, is_vowel, rafsi_tarmi,
    },
    tools::{analyze_brivla, is_brivla},
};

/// Returns the word with the given rafsi, if one exists.
pub fn search_selrafsi_from_rafsi(r: &str) -> Option<String> {
    if r != "brod" && r.len() == 4 && !r.chars().any(|c| c == '\'') {
        "aeiou"
            .chars()
            .map(|c| format!("{r}{c}"))
            .find(|gismu| RAFSI.contains_key(gismu.as_str()))
            .or_else(|| RAFSI.iter().find_map(|(v, rl)| rl.contains(&r).then(|| (*v).to_string())))
    } else {
        RAFSI.iter().find_map(|(v, rl)| rl.contains(&r).then(|| (*v).to_string()))
    }
}
/// Creates a list of selrafsi (source words) and formatted unassigned rafsi.
/// # Errors
/// None, hopefully. If you somehow do encounter one, please report it as an
/// issue in [the GitHub repository][github].
///
/// [github]: https://github.com/latkerlo/latkerlo-jvotci
#[allow(clippy::missing_panics_doc)] // .unwrap()
pub fn selrafsi_list_from_rafsi_list(
    rafsi_list: &[String],
    settings: &Settings,
) -> Result<Vec<String>, Jvonunfli> {
    let mut res = rafsi_list
        .iter()
        .map(|r| if HYPHENS.contains(&r.as_str()) { String::new() } else { r.clone() })
        .collect_vec();
    let selrafsi_list = res.iter().map(|r| search_selrafsi_from_rafsi(r)).collect_vec();
    for (i, _) in res.clone().iter().enumerate() {
        if res[i].is_empty() {
            continue;
        }
        if selrafsi_list[i].is_some() {
            res[i] = selrafsi_list[i].clone().unwrap();
        } else if rafsi_list.len() >= 2
            && i < rafsi_list.len() - 2
            && strin!(&rafsi_list[i + 1], 0) == 'y'
            && is_brivla(&format!("{}a", res[i]), &extract!(settings; y_hyphens, allow_mz))
        {
            res[i] = format!("{}-", res[i]);
        } else if is_brivla(&res[i], &extract!(settings; y_hyphens, allow_mz)) {
            // do nothing
        } else if i == rafsi_list.len() - 1
            && is_brivla(&format!("{}a", res[i]), &extract!(settings; y_hyphens, allow_mz))
        {
            res[i] = format!("{}-", res[i]);
        } else {
            res[i] = format!("-{}-", res[i]);
        }
    }
    Ok(res.iter().filter(|r| !r.is_empty()).cloned().collect_vec())
}

/// Checks if `corr` and `other` represent the same lujvo. `other` may have
/// unnecessary hyphens.
#[must_use]
pub fn compare_lujvo_pieces(corr: &[String], other: &[String]) -> bool {
    let mut i = 0;
    for part in corr {
        if part == &other[i] {
            i += 1;
            continue;
        }
        if 0 < i
            && i < other.len() - 1
            && "rn".contains(&other[i])
            && [Cvv, Cvhv].contains(&rafsi_tarmi(&other[i - 1]))
            && (i > 1 || [Ccvcv, Ccvc, Ccv].contains(&rafsi_tarmi(&other[i + 1])))
        {
            i += 1;
        }
        if part == &other[i] {
            i += 1;
        } else {
            return false;
        }
    }
    i == other.len()
}

/// Decomposes a lujvo into rafsi and hyphens.
/// # Errors
/// Any errors from [`jvokaha2`] (if the actual decomposing part fails) or
/// [`get_lujvo_from_list`] (if there are issues when re-assembling it) are
/// forwarded. A [`FakeTypeError`] is also returned if there are less than two
/// words given, and a [`DecompositionError`] is returned if the lujvo has a
/// problem that is fixable (TODO: list).
#[allow(clippy::missing_panics_doc)] // .unwrap()
pub fn jvokaha(lujvo: &str, settings: &Settings) -> Result<Vec<String>, Jvonunfli> {
    let arr = jvokaha2(lujvo, &extract!(settings; y_hyphens, allow_mz))?;
    let rafsi_tanru = arr.iter().filter(|r| r.len() > 2).map(|r| format!("-{r}-")).collect_vec();
    if rafsi_tanru.len() == 1 {
        return Err(FakeTypeError("not enough rafsi".to_string()));
    }
    let correct_lujvo = get_lujvo_from_list(&rafsi_tanru, &Settings {
        generate_cmevla: is_consonant(strin!(&arr[arr.len() - 1], -1)),
        ..extract!(settings; y_hyphens, consonants, glides, allow_mz)
    });
    if let Err(e) = correct_lujvo {
        match e {
            NoLujvoFoundError(m) => return Err(DecompositionError(m)),
            _ => return Err(e),
        }
    }
    let correct_lujvo = correct_lujvo.unwrap().0;
    let cool_and_good = if settings.y_hyphens == ForceY {
        correct_lujvo == lujvo
    } else {
        compare_lujvo_pieces(&jvokaha2(&correct_lujvo, &extract!(settings; allow_mz))?, &arr)
    };
    if cool_and_good {
        Ok(arr)
    } else {
        Err(DecompositionError(format!(
            "{{{lujvo}}} is malformed and should be {{{correct_lujvo}}}"
        )))
    }
}

/// The actual decomposing work part of [`jvokaha`].
/// # Errors
/// An [`InvalidClusterError`] is returned if the lujvo has any invalid
/// clusters.
///
/// A [`NotBrivlaError`] is returned if the lujvo starts with a CCV rafsi +
/// *'y*, because it would be a slinku'i: *uajvo* is a valid zi'evla, and we
/// want zi'evla to be able to go at the start of lujvo, so e.g. *jvo'yfliba*
/// can't be a word since you could put a *ua* in front of it to get "another"
/// lujvo.
///
/// A [`DecompositionError`] is returned if something else goes wrong. This is
/// not user-facing if you are using [`analyze_brivla`] or [`get_veljvo`].
pub fn jvokaha2(lujvo: &str, settings: &Settings) -> Result<Vec<String>, Jvonunfli> {
    let orig = lujvo;
    let mut lujvo = lujvo;
    let mut res: Vec<&str> = vec![];
    loop {
        if lujvo.is_empty() {
            return Ok(res.iter().copied().map(String::from).collect_vec());
        }
        if !res.is_empty() && res[res.len() - 1].len() != 1 {
            if strin!(lujvo, 0) == 'y'
                || settings.y_hyphens != ForceY
                    && (strsl!(lujvo, 0..2) == "nr"
                        || strin!(lujvo, 0) == 'r'
                            && lujvo.len() >= 2
                            && is_consonant(strin!(lujvo, 1)))
            {
                res.push(strsl!(lujvo, 0..1));
                lujvo = strsl!(lujvo, 1..);
                continue;
            } else if settings.y_hyphens != Standard && strsl!(lujvo, 0..2) == "'y" {
                res.push(strsl!(lujvo, 0..2));
                lujvo = strsl!(lujvo, 2..);
                continue;
            }
        }
        if rafsi_tarmi(strsl!(lujvo, 0..3)) == Cvv
            && ["ai", "ei", "oi", "au"].contains(&strsl!(lujvo, 1..3))
        {
            res.push(strsl!(lujvo, 0..3));
            lujvo = strsl!(lujvo, 3..);
            continue;
        }
        if rafsi_tarmi(strsl!(lujvo, 0..4)) == Cvhv {
            res.push(strsl!(lujvo, 0..4));
            lujvo = strsl!(lujvo, 4..);
            continue;
        }
        if [Cvcc, Ccvc].contains(&rafsi_tarmi(strsl!(lujvo, 0..4))) {
            if is_vowel(strin!(lujvo, 1)) {
                if !if settings.allow_mz { &MZ_VALID } else { &VALID }
                    .contains(&strsl!(lujvo, 2..4))
                {
                    return Err(InvalidClusterError(format!(
                        "{{{orig}}} contains an invalid cluster",
                    )));
                }
            } else if !INITIAL.contains(&strsl!(lujvo, 0..2)) {
                return Err(InvalidClusterError(format!(
                    "{{{orig}}} starts with an invalid cluster",
                )));
            }
            if lujvo.len() == 4 || strin!(lujvo, 4) == 'y' {
                res.push(strsl!(lujvo, 0..4));
                if strin!(lujvo, 4) == 'y' {
                    res.push("y");
                }
                lujvo = strsl!(lujvo, 5..);
                continue;
            }
        }
        if [Cvccv, Ccvcv].contains(&rafsi_tarmi(lujvo)) {
            res.push(lujvo);
            return Ok(res.iter().copied().map(String::from).collect_vec());
        }
        if rafsi_tarmi(strsl!(lujvo, 0..3)) == Cvc {
            res.push(strsl!(lujvo, 0..3));
            lujvo = strsl!(lujvo, 3..);
            continue;
        }
        if rafsi_tarmi(strsl!(lujvo, 0..3)) == Ccv {
            if !INITIAL.contains(&strsl!(lujvo, 0..2)) {
                return Err(InvalidClusterError(format!(
                    "{{{orig}}} starts with an invalid cluster",
                )));
            }
            if lujvo == orig && strsl!(lujvo, 3..5) == "'y" {
                return Err(NotBrivlaError(format!(
                    "{{{orig}}} starts with CCV'y, making it a slinku'i"
                )));
            }
            res.push(strsl!(lujvo, 0..3));
            lujvo = strsl!(lujvo, 3..);
            continue;
        }
        return Err(DecompositionError(format!("{{{orig}}} can't be decomposed")));
    }
}

/// Calculates the score for a lujvo.
/// # Errors
/// Errors are forwarded from [`get_veljvo`] and [`analyze_brivla`].
pub fn score_lujvo(lujvo: &str, settings: &Settings) -> Result<i32, Jvonunfli> {
    get_veljvo(lujvo, settings)?;
    let decomp = analyze_brivla(lujvo, settings)?.1;
    Ok(decomp
        .iter()
        .map(|r| {
            if ["y", "n", "r", ""].contains(&r.as_str()) { 1100 * r.len() as i32 } else { score(r) }
        })
        .sum::<i32>()
        - tiebreak(lujvo))
}

/// Gets the selrafsi (source tanru) and formatted unassigned rafsi for this
/// lujvo.
/// # Errors
/// Errors are forwarded from [`analyze_brivla`]. Additionally a
/// [`DecompositionError`] is returned if given something other than a lujvo.
pub fn get_veljvo(lujvo: &str, settings: &Settings) -> Result<Vec<String>, Jvonunfli> {
    let (b_type, rafsi_list) = analyze_brivla(
        lujvo,
        &extract!(
            settings;
            y_hyphens,
            exp_rafsi,
            consonants,
            generate_cmevla,
            allow_mz
        ),
    )?;
    if ![Lujvo, ExtendedLujvo, Cmevla].contains(&b_type) {
        return Err(DecompositionError(format!(
            "{{{lujvo}}} is a {}, not a lujvo or decomposable cmevla",
            b_type.to_string().to_lowercase()
        )));
    }
    selrafsi_list_from_rafsi_list(&rafsi_list, &extract!(settings; y_hyphens, allow_mz))
}

File: py/latkerlo_jvotci/katna.py
"""
Copyright (c) 2021 sozysozbot (https://github.com/sozysozbot)
Licensed under the MIT License

Modified by latkerlo (https://github.com/latkerlo), Copyright (c) 2023-2024
"""

from latkerlo_jvotci.tarmi import *
from latkerlo_jvotci.tools import is_brivla, analyse_brivla
from latkerlo_jvotci.jvozba import get_lujvo_from_list
from latkerlo_jvotci.exceptions import DecompositionError, InvalidClusterError, NotBrivlaError, NoLujvoFoundError
from latkerlo_jvotci.rafsi import RAFSI_LIST


def search_selrafsi_from_rafsi(rafsi: str) -> str:
    """
    Return the selrafsi for a given rafsi, if one exists.
    Otherwise, None is returned.

    :param rafsi: The rafsi to search for.
    :return: The corresponding selrafsi, if applicable, otherwise None.
    """
    if rafsi != "brod" and len(rafsi) == 4 and "'" not in rafsi:  # 4-letter rafsi
        for u in range(5):
            gismu_candid = rafsi + "aeiou"[u]
            if gismu_candid in RAFSI_LIST:
                return gismu_candid
    for valsi, rafsi_list in RAFSI_LIST.items():
        if rafsi in rafsi_list:
            return valsi


def selrafsi_list_from_rafsi_list(rafsi_list: list[str], allow_mz=False) -> list[str]:
    """
    Create a list of selrafsi and formatted rafsi from a list of rafsi.

    Example:
    ["lat", "mot", "kelr", "y", "kerlo"] ->
    ["mlatu", "-mot-", "kelr-", "kerlo"]

    :param rafsi_list: List of rafsi and hyphens (a decomposed word).
    :param allow_mz: True if mz is a valid consonant cluster.
    :return: List of selrafsi and formatted rafsi.
    """
    result = [("" if x in HYPHENS else x) for x in rafsi_list]
    selrafsi_list = [search_selrafsi_from_rafsi(x) for x in result]
    for i, selrafsi in enumerate(selrafsi_list):
        if not result[i]:
            continue

        if selrafsi is not None:
            result[i] = selrafsi
        elif i < len(rafsi_list) - 2 and rafsi_list[i+1][0] == "y" and is_brivla(result[i] + "a", allow_mz=allow_mz):
            result[i] = f'{result[i]}-'
        elif is_brivla(result[i], allow_mz=allow_mz):
            pass
        elif i == len(rafsi_list) - 1 and is_brivla(result[i] + "a", allow_mz=allow_mz):
            result[i] = f'{result[i]}-'
        else:
            result[i] = f'-{result[i]}-'
    return [x for x in result if len(x) > 0]


# Assumes... something? TODO
def compare_lujvo_pieces(corr: [str], other: [str]) -> bool:
    """
    Check if corr and other represent the same lujvo.
    other may have unnecessary hyphens.

    :param corr: A list of parts of the correct lujvo.
    :param other: A list of parts of a candidate to test.
    :return: True if the lujvo are the same except for hyphens.
    """
    i = 0
    for j, part in enumerate(corr):
        if part == other[i]:
            i += 1
            continue

        if (
            0 < i < len(other) - 1 and
            other[i] in ["r", "n"] and
            rafsi_tarmi(other[i-1]) in [CVV, CVhV] and
            (i > 1 or rafsi_tarmi(other[i+1]) in [CCVCV, CCVC, CCV])
        ):
            i += 1

        if part == other[i]:
            i += 1
        else:
            return False
    return i == len(other)


# jvokaha("fu'ivla") --> ["fu'i", "vla"]
# jvokaha("fu'irvla") --> error // because r-hyphen is unnecessary
# jvokaha("pasymabru") --> ["pas", "y", "mabru"]
# jvokaha("pasmabru") --> error // because {pasmabru} is actually {pa smabru}
def jvokaha(
        lujvo: str,
        allow_rn_hyphens: bool = False,
        y_hyphens: str = STANDARD,
        consonants: str = CLUSTER,
        glides: bool = False,
        allow_mz: bool = False) \
        -> list[str]:
    """
    Decompose lujvo to get a list of pieces (rafsi and hyphens). Raise
    an error if the lujvo is not well-formed.

    :param lujvo: A lujvo to decompose.
    :param allow_rn_hyphens: True if unnecessary r & n hyphens are
    allowed.
    :param y_hyphens: Which y-hyphen rules to use.
    :param consonants: Which consonant rules to use.
    :param glides: True if glides count as consonants.
    :param allow_mz: True if mz is a valid consonant cluster.
    :return: List of lujvo pieces (rafsi and hyphens).
    """
    arr = jvokaha_2(lujvo, y_hyphens=y_hyphens, allow_mz=allow_mz)

    rafsi_tanru = [f'-{x}-' for x in arr if len(x) > 2]
    try:
        correct_lujvo = get_lujvo_from_list(
            rafsi_tanru,
            generate_cmevla=is_consonant(arr[-1][-1]),
            y_hyphens=y_hyphens,
            consonants=consonants,
            glides=glides,
            allow_mz=allow_mz)[0]
    except NoLujvoFoundError:
        raise DecompositionError(f"no lujvo for {rafsi_tanru}")

    if allow_rn_hyphens and not y_hyphens == FORCE_Y:
        cool_and_good = compare_lujvo_pieces(jvokaha_2(correct_lujvo, allow_mz=allow_mz), arr)
    else:
        cool_and_good = correct_lujvo == lujvo

    if cool_and_good:
        return arr
    else:
        raise DecompositionError("malformed lujvo {" + lujvo +
                                 "}; it should be {" + correct_lujvo + "}")


# jvokaha2("fu'ivla") --> ["fu'i", "vla"]
# jvokaha2("fu'irvla") --> ["fu'i", "r", "vla"]
# jvokaha2("pasymabru") --> ["pas", "y", "mabru"]
# jvokaha2("pasmabru") --> ["pas", "mabru"]
def jvokaha_2(
        lujvo: str,
        y_hyphens: str = STANDARD,
        allow_mz: bool = False) \
        -> list[str]:
    """
    Decompose lujvo to get a list of pieces (rafsi and hyphens).
    Raises an error if the string is not decomposable, but not if it
    is invalid for other reasons.

    :param lujvo: A lujvo to decompose.
    :param y_hyphens: Which y-hyphen rules to use.
    :param allow_mz: True if mz is a valid consonant cluster.
    :return: List of lujvo pieces (rafsi and hyphens).
    """
    original_lujvo = lujvo
    res = []
    while True:
        if lujvo == "":
            return res

        # remove hyphen
        if len(res) > 0 and len(res[-1]) != 1:  # hyphen cannot begin a word; nor can two hyphens
            if lujvo[0] == "y":  # y-hyphen
                res.append(lujvo[0])
                lujvo = lujvo[1:]
                continue
            elif y_hyphens != FORCE_Y and (
                lujvo[:2] == "nr"  # n-hyphen is only allowed before r
                or lujvo[0] == "r" and is_consonant(lujvo[1])  # r followed by a consonant
            ):
                res.append(lujvo[0])
                lujvo = lujvo[1:]
                continue
            elif y_hyphens != STANDARD and lujvo[:2] == "'y":
                res.append(lujvo[:2])
                lujvo = lujvo[2:]
                continue

        # drop rafsi from front

        # CVV can always be dropped
        if rafsi_tarmi(lujvo[:3]) == CVV and lujvo[1:3] in ["ai", "ei", "oi", "au"]:
            res.append(lujvo[:3])
            lujvo = lujvo[3:]
            continue

        # CV'V can always be dropped
        if rafsi_tarmi(lujvo[:4]) == CVhV:
            res.append(lujvo[:4])
            lujvo = lujvo[4:]
            continue

        # CVCCY and CCVCY can always be dropped
        if rafsi_tarmi(lujvo[:4]) in [CVCC, CCVC]:
            if is_vowel(lujvo[1]):
                if lujvo[2:4] not in (MZ_VALID if allow_mz else VALID):
                    raise InvalidClusterError(f"Invalid cluster {{{lujvo[2:4]}}} in {{{original_lujvo}}}")
            else:
                if lujvo[0:2] not in INITIAL:
                    raise InvalidClusterError(f"Invalid initial cluster {{{lujvo[0:2]}}} in {{{original_lujvo}}}")

            if len(lujvo) == 4 or lujvo[4] == "y":
                res.append(lujvo[:4])
                try:
                    if lujvo[4] == "y":
                        res.append("y")
                except IndexError:
                    pass
                lujvo = lujvo[5:]
                continue

        # the final rafsi can be 5-letter
        if rafsi_tarmi(lujvo) in [CVCCV, CCVCV]:
            res.append(lujvo)
            return res

        if rafsi_tarmi(lujvo[:3]) == CVC:
            res.append(lujvo[:3])
            lujvo = lujvo[3:]
            continue

        if rafsi_tarmi(lujvo[:3]) == CCV:
            if lujvo[0:2] not in INITIAL:
                raise InvalidClusterError(f"Invalid initial cluster {{{lujvo[0:2]}}} in {{{original_lujvo}}}")
            res.append(lujvo[:3])
            lujvo = lujvo[3:]
            continue

        # if all fails...
        raise DecompositionError("Failed to decompose {" + original_lujvo + "}")


def get_veljvo(
        lujvo: str,
        y_hyphens: str = STANDARD,
        exp_rafsi_shapes: bool = False,
        consonants: str = CLUSTER,
        glides: bool = False,
        allow_mz: bool = False) \
        -> list[str]:
    """
    Decompose a lujvo into a list of selrafsi and formatted rafsi.

    :param lujvo: Lujvo to decompose.
    :param y_hyphens: Which y-hyphen rules to use.
    :param exp_rafsi_shapes: True if experimental rafsi shapes are allowed.
    :param consonants: Which consonant rules to use.
    :param glides: True if glides count as consonants.
    :param allow_mz: True if mz is a valid consonant cluster.
    :return: List of selrafsi and rafsi.
    """
    b_type, rafsi_list = analyse_brivla(
        lujvo,
        y_hyphens=y_hyphens,
        exp_rafsi_shapes=exp_rafsi_shapes,
        consonants=consonants,
        glides=glides,
        allow_mz=allow_mz
    )

    if b_type not in [LUJVO, EXTENDED_LUJVO, CMEVLA]:
        raise DecompositionError(f"Valsi is of type {b_type}")
    return selrafsi_list_from_rafsi_list(rafsi_list, allow_mz)


if __name__ == "__main__":
    # Terrible CLI for quick testing. Use at own risk.
    y_setting = STANDARD
    raf_setting = False
    con_setting = CLUSTER
    glide_setting = False
    mz_setting = False
    while True:
        string = input(f"({y_setting[0]}{str(raf_setting)[0]}{con_setting[0]}{str(glide_setting)[0]}{str(mz_setting)[0]}) Enter a lujvo: ")
        if not string:
            break
        if string[0] == "/":
            for change in string[1:].lower().split():
                match change:
                    case "standard":
                        y_setting = STANDARD
                    case "allow":
                        y_setting = ALLOW_Y
                    case "force":
                        y_setting = FORCE_Y
                    case "more":
                        raf_setting = True
                    case "less":
                        raf_setting = False
                    case "cluster":
                        con_setting = CLUSTER
                    case "2c":
                        con_setting = TWO_CONSONANTS
                    case "1c":
                        con_setting = ONE_CONSONANT
                    case "glide":
                        glide_setting = True
                    case "glidnt":
                        glide_setting = False
                    case "mz":
                        mz_setting = True
                    case "mznt":
                        mz_setting = False
                    case _:
                        continue
            continue
        try:
            print(analyse_brivla(string, y_hyphens=y_setting, exp_rafsi_shapes=raf_setting, consonants=con_setting, glides=glide_setting, allow_mz=mz_setting))
            print(get_veljvo(string, y_hyphens=y_setting, exp_rafsi_shapes=raf_setting, consonants=con_setting, glides=glide_setting, allow_mz=mz_setting))
        except (DecompositionError, NotBrivlaError) as e:
            print(f"Error: {e}")
